### PR TITLE
Fix #6007: Get the classloader that loaded the application

### LINKED ISSUE #6007: Calling `quoted.Expr.{run,show}` crashes at runtime in Dotty 0.13.0-RC1, but not 0.12.0
When testing if #5997 was reproducible on the latest version of Dotty, I found instead that trying to run my example on version 0.13.0-RC1 crashes at runtime. The same code running on 0.12.0 or 0.12.0-RC1 does not crash.

Steps to reproduce (on Ubuntu 16.04, with sbt 1.2.7):
```
$ git clone git@github.com:lampepfl/dotty-example-project.git
$ cd dotty-example-project
$ git log
commit a753b14e281bbaa6c69f26298913ad6feba969c7
Merge: 3169229 34953cd
Author: Aggelos Biboudis <aggelos.biboudis@epfl.ch>
Date:   Tue Feb 26 19:34:33 2019 +0100

    Merge pull request #25 from lampepfl/dotty-0.13
    
    Release Dotty 0.13.0-RC1
[rest of log omitted]
$ sbt
sbt:dotty-example-project> about
[info] This is sbt 1.2.7
[info] The current project is ProjectRef(uri("file:/.../dotty-example-project/"), "root") 0.1.0
[info] The current project is built against Scala 0.13.0-RC1
[info] Available Plugins
[info]  - dotty.tools.sbtplugin.DottyIDEPlugin
[info]  - dotty.tools.sbtplugin.DottyPlugin
[info]  - sbt.ScriptedPlugin
[info]  - sbt.plugins.CorePlugin
[info]  - sbt.plugins.Giter8TemplatePlugin
[info]  - sbt.plugins.IvyPlugin
[info]  - sbt.plugins.JUnitXmlReportPlugin
[info]  - sbt.plugins.JvmPlugin
[info]  - sbt.plugins.SbtPlugin
[info] sbt, sbt plugins, and build definitions are using Scala 2.12.7
sbt:dotty-example-project> console
scala> import quoted.Toolbox.Default._
scala> val v = '{ (if true then Some(1) else None).map(v => v+1) }
scala> v.show
scala.quoted.Toolbox$ToolboxNotFoundException: Could not load the Toolbox class `dotty.tools.dotc.quoted.ToolboxImpl` from the JVM classpath. Make sure that the compiler is on the JVM classpath.
	at scala.quoted.Toolbox$.make(Toolbox.scala:29)
	at scala.quoted.Toolbox$Default$.make(Toolbox.scala:15)
	at rs$line$4$.<init>(rs$line$4:1)
[huge stacktrace omitted]
Caused by: java.lang.ClassNotFoundException: dotty.tools.dotc.quoted.ToolboxImpl
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at scala.quoted.Toolbox$.make(Toolbox.scala:21)
	... 54 more
scala> v.run
[the same or very similar error]
```

Let me know if you need any other information in order to reproduce the issue. I abridged the stacktraces for readability, but perhaps I cut out something important.

At first glance, it seems like either the Dotty sbt plugin is misbehaving by somehow not setting the classpath properly, or Dotty is doing something very weird. I don't know enough about the internals to determine which.

### PR DESCRIPTION
**Pull request by biboudis**
The issue is that the classloader used to load the library is not the same as the one that loads the rest of the aplication. The toolbox requires the classloader that loaded the main of the application, the one that has all the classpath dependencies of the appliaction.

### CODE REVIEW

**Comment by biboudis** [comment]
Added a scripted test to emulate #6007

`> sbt-dotty/scripted sbt-dotty/quoted-example-project`

File: library/src/scala/quoted/Toolbox.scala [hunk 1]
```diff
@@ -16,12 +16,35 @@ object Toolbox {
...
 
-  def make(implicit settings: Settings): Toolbox = {
-    val cl = getClass.getClassLoader
+  /** Create a new instance of the toolbox.
+    *
+    * This instance of the toolbox tries to recover the classloader of the application based on
+    * the classloader of the class that call make. This may not always be the correct classloader,
+    * in which case the classloader must be passed explicitly.
+    *
+    * @param settings TODO
```
Comment (biboudis) [inline library/src/scala/quoted/Toolbox.scala]: TODO

File: sbt-dotty/sbt-test/sbt-dotty/quoted-example-project/build.sbt [hunk 2]
```diff
@@ -0,0 +1,7 @@
+scalaVersion := sys.props("plugin.scalaVersion")
+
+libraryDependencies ++= Seq(
+      "ch.epfl.lamp" % "dotty_0.14" % scalaVersion.value,
+      "ch.epfl.lamp" % "dotty_0.14" % scalaVersion.value % "test->runtime",
```
Comment (allanrenucci) [inline sbt-dotty/sbt-test/sbt-dotty/quoted-example-project/build.sbt]: The dotty version should not be hardcoded

File: library/src/scala/quoted/Toolbox.scala [hunk 3]
```diff
@@ -16,12 +16,35 @@ object Toolbox {
...
+  /** Create a new instance of the toolbox.
+    *
+    * This instance of the toolbox tries to recover the classloader of the application based on
+    * the classloader of the class that call make. This may not always be the correct classloader,
+    * in which case the classloader must be passed explicitly.
+    *
+    * @param settings toolbox settings
+    * @return A new instance of the toolbox
+    */
+  @forceInline def make(implicit settings: Settings): Toolbox = {
```
Comment (nicolasstucki) [inline library/src/scala/quoted/Toolbox.scala]: @smarter could you have a look at this method?

File: sbt-dotty/sbt-test/sbt-dotty/quoted-example-project/build.sbt (hunk 2 above)
Comment (smarter) [reply sbt-dotty/sbt-test/sbt-dotty/quoted-example-project/build.sbt]: Also use "dotty-compiler", not "dotty".

File: library/src/scala/quoted/Toolbox.scala (hunk 3 above)
Comment (smarter) [reply library/src/scala/quoted/Toolbox.scala]: Looks super hacky to me. I'd like to know what scalac does.

Comment (nicolasstucki) [reply library/src/scala/quoted/Toolbox.scala]: They bailed out on trying to recover it. They basically have the equivalent of the other `make` that asks for the class loader explicitly. In the docs they just tell the user to pass `getClass.classLoader`

Comment (nicolasstucki) [reply library/src/scala/quoted/Toolbox.scala]: I will remove this variant for now.

File: sbt-dotty/sbt-test/sbt-dotty/quoted-example-project/build.sbt (hunk 2 above)
Comment (nicolasstucki) [reply sbt-dotty/sbt-test/sbt-dotty/quoted-example-project/build.sbt]: Opened #6207 to for this issue

File: library/src/scala/quoted/Toolbox.scala (hunk 3 above)
Comment (nicolasstucki) [reply library/src/scala/quoted/Toolbox.scala]: This hacky method was removed

**Pull request merged**